staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -72,6 +72,59 @@ func init() {
 	dbMetricsMonitors = make(map[string]struct{})
 }
 
+type activeETCD3Client struct {
+	client  *clientv3.Client
+	destroy DestroyFunc
+	refs    int
+}
+
+type etcd3ClientCache struct {
+	mx     sync.Mutex
+	active map[string]*activeETCD3Client
+}
+
+func newETCD3ClientCache() *etcd3ClientCache {
+	return &etcd3ClientCache{active: make(map[string]*activeETCD3Client)}
+}
+
+// For returns a compacted etcd v3 client for the supplied transport config. One
+// client and compactor are started for each transport. Callers must call the
+// returned destroy function once they are done with the client.
+func (e *etcd3ClientCache) For(tc storagebackend.TransportConfig, compactionInterval time.Duration) (*clientv3.Client, DestroyFunc, error) {
+	e.mx.Lock()
+	defer e.mx.Unlock()
+
+	k := fmt.Sprintf("%v", tc) // gives: {[server1 server2] keyFile certFile caFile}
+	if a, ok := e.active[k]; ok {
+		a.refs++
+		return a.client, a.destroy, nil
+	}
+
+	c, err := newETCD3Client(tc)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot create new etcd3 client: %w", err)
+	}
+
+	// TODO(negz): Handle requests for a client for the same transport but with
+	// more frequent compaction.
+	ctx, cancel := context.WithCancel(context.Background())
+	etcd3.StartCompactor(ctx, c, compactionInterval)
+
+	d := func() {
+		e.mx.Lock()
+		defer e.mx.Unlock()
+		e.active[k].refs--
+		if e.active[k].refs <= 0 {
+			cancel()
```

Comment:
if we go with this I think that this cache operations should have their own methods 
```
e.Remove(client string)
e.Get(client string) (*clientv3.Client, DestroyFunc, error)
e.Add(client string, c *clientv3.Client, d DestroyFunc, ref int)
```

Reply:
Agreed - I almost built it this way to start with. Will update.

Reply:
Done!